### PR TITLE
Total Uptime provider for libdns - resolves #90

### LINKED ISSUE #90: New provider: Total Uptime
Hello, I'd like to build an integration with one of our providers Total Uptime.
Thanks!

### PR DESCRIPTION
**Pull request by kmei3560**
Hello, this PR is an implementation of libdns for Total Uptime DNS provider.

Please let me know if I've not submitted this correctly. Also, feel free to give any feedback on the code itself, or let me know if I've done something wrong.

Thank you

### CODE REVIEW

File: totaluptime/README.md [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+	"fmt"
+
+	"github.com/libdns/totaluptime"
+)
+
+func main() {
+	var provider totaluptime.Provider
+
+	provider.Username = USERNAME      // provider API username
+	provider.Password = PASSWORD      // provider API password
```
Comment (mholt) [inline totaluptime/README.md:45]:
```suggestion
	provider := totaluptime.Provider{
		provider.Username = "USERNAME",      // provider API username
		provider.Password = "PASSWORD",      // provider API password
	}
```

I think this is more typical convention

File: totaluptime/examples/go.mod [hunk 2]
```diff
@@ -0,0 +1,9 @@
+module github.com/kmei3560/libdns/totaluptime/examples
```
Comment (mholt) [inline totaluptime/examples/go.mod:1]: Don't forget to change this when you push to the libdns repo

**Review by mholt — COMMENTED** [review]
Thanks for the submission!

Overall pretty good. I would recommend getting rid of the Print lines that aren't in the example file; generally you should return the first error right away so that they can be fixed.

Let me know when you're ready and I'll make the repos for you to push to. Thanks!

File: totaluptime/README.md (hunk 1 above)
Comment (kmei3560) [reply totaluptime/README.md:45]: Got it. Thanks

**Comment by kmei3560** [comment]
Can you please take another look when you have a chance?

Also, I don't understand what you mean by this statement:

> I would recommend getting rid of the Print lines that aren't in the example file; generally you should return the first error right away so that they can be fixed.

Thanks

**Comment by kmei3560** [comment]
Please review latest change and see if this is better (I agree it's more proper with `log.Fatalln`)
If everything looks good, you can go ahead and setup the new repo. Thanks!

File: totaluptime/provider.go [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+		return nil, fmt.Errorf("record append cannot proceed: unknown zone: %s", zone)
+	}
+
+	// append each record individually
+	for _, rec := range records {
+		// identify record type
+		typeName, err := convertRecordTypeToProvider(rec.Type)
+		if err != nil {
+			log.Printf("Unable to identify record type; skipping: %s: %s\n", rec.Type, rec.Name)
+			continue // skip to next record
```
Comment (mholt) [inline totaluptime/provider.go:158]:
```suggestion
			return nil, err
```

There's a lot of lines like this... ideally, don't emit logs -- just do, or do not. (There is no try.)

I'm not familiar with TU, so it's up to you decide: if this case is unexpected, then return the error. If it's normal and harmless (doesn't affect the final result; i.e. this is a debug-level log), then remove the log and just continue.

**Review by mholt — COMMENTED** [review]
(Sorry, been catching up after the holiday, and now taking care of sick kids.)

Hopefully this clarifies things! Basically, try to figure out of each of these cases is normal and OK or not. If not, return an error. If so, don't print a log.

**Comment by kmei3560** [comment]
Gotcha. It makes a lot of sense to me when you say "don't emit logs"... yes I tend to be overly verbose with logging.
I tried to eliminate the scenarios you mentioned (DO or DONT, but don't emit logs unnecessarily)

There may be a few more cases remaining, so just let me know if I haven't taken this far enough. Typically I like to log a condition that caused something unexpected. So for example, if an incorrect domain was requested to be deleted - I would want to log that the domain could not be found (to help user understand).

Thank you very much for your Go wisdom!  Very much appreciated! 
Also, sorry to hear about sick kids. Hope everyone feels better soon!

**Review by mholt — APPROVED** [review]
Great, looks better now :) Thanks for the contribution! I'll create the repos and add you to them. (libdns and caddy-dns as well)

**Comment by kmei3560** [comment]
Sounds good - and thank you!  For the record, I do not use `caddy` as-of-yet. I currently only use `certmagic` library and `libdns` providers.

**Comment by mholt** [comment]
Ah, apologies for presuming. (You're probably the first who contributed a provider but isn't using Caddy.) I'll leave the repo and invite there, but don't feel obligated!

**Pull request closed without merging**